docs/dev: introduce "Distribution Point" concept

sgotti opened the pull request:
A Distribution Point represents the method to fetch an image starting from an
input string.

Before this concept was introduced, rkt used the ImageType concept, the
ImageType was mapped to the input image string format and internally
covered multiple concepts like distribution, transport and image type
(hidden since all are now appc ACIs).

The distribution point concept will be used as the primary information in the
different layers of rkt (fetching but also for references in a CAS/ref store).

This patch introduces the distribution point concept. See
Documentation/devel/distribution_point.md for a detailed description.

lucab commented:
/cc @dgonyeo (this is part of the store rework I mentioned)

lucab commented:
Can we please put the PR details inside a markdown doc? Even if not final, it will make discussion easier.

> Should the archive URI be split in one distribution type per format? Something like aciarchive ociimagelayoutarchive

Bikeshed level: I think so. I also think schemes should be scoped (as they are rkt specific) and be made future-proof. Something like:
 * `rkt:appc:v=0:name...`
 * `rkt:docker:v=0:name...`
 * `rkt:aci-archive:v=0:ArchiveUrl...`
 * `rkt:oci-archive:v=0:ArchiveUrl...`

Where we basically reserve for ourself a single scheme (`rkt:`) and we internally manage sub-schemes (in the form of `format:k=v:`).

sgotti commented:
> Can we please put the PR details inside a markdown doc? Even if not final, it will make discussion easier.

Done, I put the details in a not final doc in Documentation/devel/distribution.md



> Bikeshed level: I think so. I also think schemes should be scoped (as they are rkt specific) and be made future-proof. Something like:
> 
> rkt:appc:v=0:name...
> rkt:docker:v=0:name...
> rkt:aci-archive:v=0:ArchiveUrl...
> rkt:oci-archive:v=0:ArchiveUrl...
> Where we basically reserve for ourself a single scheme (rkt:) and we internally manage sub-schemes (in the form of format:k=v:).

This seems good. Version will be the sub-scheme version right? (to be able to update its internal format without putting the version directly in the subscheme and changing the subscheme name).
Should `v=...` be mandatory for every subscheme and with a fixed position (after the subscheme) or just a possible subscheme dependent `k=v` entry? If mandatory can we just use semantic versioning without a key=value for it like `rkt:appc:v0.1:....`?

@s-urbaniak? thoughts on the URI (as names) direction and the suggested formats?

lucab commented:
> Version will be the sub-scheme version right? (to be able to update its internal format without putting the version directly in the subscheme and changing the subscheme name).

Right (and yes).

> Should v=... be mandatory for every subscheme and with a fixed position (after the subscheme) or just a possible subscheme dependent k=v entry?

The former (which in case can be expanded into the latter given a k=v array, but I prefer not going that route if there's no need to)

> If mandatory can we just use semantic versioning without a key=value for it like rkt:appc:v0.1:....?

I would avoid complexity and just keep a `uint`. This is mostly an internal format and we stick a version there "just in case".

sgotti commented:
> Bikeshed level: I think so. I also think schemes should be scoped (as they are rkt specific)

Since my hope is that all this effort and others to decouple different layers will be shared between multiple tools (acbuild, new tools etc...) I won't make them too much rkt specific. If we want to go with an unique scheme I'll try to find a more generic name ("distribution" ?).

lucab commented:
> Since my hope is that all this effort and others to decouple different layers will be shared between multiple tools (acbuild, new tools etc...) I won't make them too much rkt specific. If we want to go with an unique scheme I'll try to find a more generic name ("distribution" ?).

So the owner of this would be some kind of `pkg/container-distribution` and not rkt itself, right? If so, just reflect it in some mnemonic way in the scheme. Maybe `codi:` or similar, your choice. My concerns are just:
 * we should pick a single scheme-root that we can reserve for ourself
 * we should stop squatting somebody else generic scheme (like `appc:` or `docker:)`

(I hope we are not starting to over-design here)

dgonyeo commented:
> Since my hope is that all this effort and others to decouple different layers will be shared between multiple tools (acbuild, new tools etc...) I won't make them too much rkt specific.

It would be ideal to share this with acbuild at the very least. Is there a difference between distributions here and distribution handlers in https://github.com/coreos/rkt/pull/2964? If not, it sounded like distribution handlers are responsible for interacting with rkt's store, which means that if an external project wants to use these distribution handlers it would also have to use rkt's store logic.

> I'm not sure if distribution is the best name for this concept

I think distribution is fine.

sgotti commented:
> ? If not, it sounded like distribution handlers are responsible for interacting with rkt's store, which means that if an external project wants to use these distribution handlers it would also have to use rkt's store logic.

That's a good point.
I'm not sure how it could be decoupled from the store:

* for example in oci the distribution should check in the store if a blob already exists or it needs to be fetched. Doing a fetching phase without the store will mean fetching all blobs also if already available. 
* In #2964 also the transport cache should interact with the store to see if the blob for the specified transport url exists.

One idea will be to define a (read-write) store interface with the required CAS and ref methods (WriteBlob, ReadBlob, ListBlobs, HasBlob, SetRef ...). I have to experiment a little bit more with the store refactor to see if this will work or if some image type differences (ACI, OCI image) will force us to use dedicated store functions per image type (some example can be handling ACI manifest cache to avoid extracting the manifest every time its needed from an image),

sgotti commented:
Updated with the proposed changes:

* A unique global scheme (now `cimd`... Please find a better name that recalls `container` `image` and `distribution` :sweat_smile: )
* Moved archive to `aci-archive` (no more FORMAT parameter). In future there'll be an `oci-image-layout-archive`.
* Removed SimpleString() from the distribution interface since it's better leaving this to the distribution user (there may been different ways for a simple string) like is done for the conversion from a simple string to the distribution uri. The docker distribution provides a SimpleDockerString that generates what an user will expect (removing default index, default repo when on default index and tag if latest, see the tests).
* More tests.

sgotti commented:
I thinks its ready for review (will adapt the docs if the naming are settled). PTAL.

The unique change is, to support additional formats with simple string paths (OCI image layout archives, directory based layouts) that an ACI archive expressed with a simple string requires a .aci extension.

dgonyeo commented on common/distribution/aciarchive.go:
```diff
@@ -0,0 +1,99 @@
+// Examples:
+// cimd:aci-archive:v=0:file%3A%2F%2Fabsolute%2Fpath%2Fto%2Ffile
+// cimd:aci-archive:v=0:https%3A%2F%2Fexample.com%2Fapp.aci
+type ACIArchive struct {
```
Maybe this should have a transport URL for the image and a transport URL for the image's signature?

The signature URL can't necessarily be derived from the image URL.

sgotti replied:
Good question. I had the same though when designing it. Looking at rkt you can pass the image signature with `--signature` as an additional option for http, file etc... and this can also be used to override the appc discovered signature.

So I thought to leave it outside distribution because I'd also have to find a way to specify an ovverride signature for appc distribution and that it will generate different distributions for every different passed signature.

I'm not completely sure that my reasoning is correct so I'd like to hear you thoughts!

sgotti commented:
Updated adding some concepts around distribution types (Name based and Location based distributions). These can become useful to understand better the proposal and for future work on a reference(distribution) based image handling logic.

@philips @jonboulle Since you're part of the OCI image spec maintainers, I'd really like to hear your thoughs on this:
* Does it fits the future work on OCI image additional distributions (outside the current docker registry distribution).
* I digged through the various oci image issues and mails but I had hard times trying to find a final decision on what a distributions is and how its related to image "name" etc... Do you think that this could also be helpful for the OCI image spec to define what a distribution is?

@yifan-gu @euank I'd really like to hear your thoughs on this:
* Do you think that this will be helpful for future k8s work on supporting multiple image types. Since more than image types I really think k8s should support multiple image distributions (probably only Name based distributions)?

lucab commented:
@sgotti I saw that docs and code here is slightly overlapping-but-different than https://github.com/coreos/rkt/pull/3071. I haven't had time yet for a final pass, how do we want to proceed here? Maybe we declare the design doc stable (if you feel it), we merge it and then we follow with the implementation piece by piece?

sgotti commented:
@lucab Yeah, #3071 is based on this (and also uses the new treestore in #2972) so it may be left behind with older commits, but this is the official one.

I'm ok splitting the design/proposal from the implementation. Before merging, it'll be probably better to hear other thoughts since this is a change that impacts (and hopefully helps) different parts of rkt and related projects.

yifan-gu commented on Documentation/devel/distribution.md:
```diff
@@ -0,0 +1,145 @@
+|----------------|----------|---------------------------------------------------------------------------|------------------|
+| Appc           | Name     | `cimd:appc:v=0:name?label01=....&label02=...`                             | ACIArchive       |
+| Docker         | Name     | `cimd:docker:v=0:[REGISTRY_HOST[:REGISTRY_PORT]/]NAME[:TAG&#124;@DIGEST]` | <none>           |
+| ACIArchive     | Location | `cimd:aci-archive:v=0:ArchiveURL?query...`                                |                  |
```
I don't get why we need `Location` type? What's the benefits we get compared to just treating them as a "discovery string"

sgotti replied:
I'm not sure I completely understand your question. I'll add more contents but please let me know if there's something not clear or if you don't agree on something.

* The distribution is only defined by its URI.
* The distribution Type is only a logical distinction but I think it's very important to clarify this distinction and its role since it'll affect the logic of the various distribution related implementations (like #2964 #3071). I tried to explain it [here](https://github.com/coreos/rkt/pull/2953/files/50c7a2f3072f67bb6254e1a09270e6e52ac168d9#diff-28042cdb928d9fb5227b3ced90a10356R13).

I'll try to explain my reasoning behind this:
* Name based distribution should be the primary way to fetch/reference an image.
 * for example appc ACI dependencies can be resolved only by a name based distribution (appc).
* Location based distribution should not be directly used because they map a fixed location to an image but they exists and are needed for different reason:
 * rkt provides a way to fetch and run an image by a location (the current file or http uri are an aci-archive distribution)
 * Some name based distribution will naturally resolve to a location based distribution:
  * appc distribution will do discovery and the discovered url is a location based distribution (aci-archive)
  * the same will probably apply for OCI image (from a name to an oci-image-layout location distribution)
 * in a reference based store logic (see the notes in https://github.com/coreos/rkt/pull/3071/commits/23313af1c3dac2fb24fe41f9a7c5eaca573e45dd#diff-146353160c5a20311afdb1ccda76b178R299) it could be useful to let the user map a location based distribution to a name based one just because a name based is needed for appc distribution but I'd like to "import" an image directly by a URL. Instead it doesn't makes sense to map a name to a location or a location to a location distribution.

I intentionally avoided the term "discovery" because:
* it's just a way to implement a name based distribution
* I don't consider a location distribution related with discovery, you already provide the final location where the image is contained (aci-archive, oci-image-layout).

sgotti replied:
updated the above comment to fix some errors!

yifan-gu replied:
I can't conceptually see the big difference between a final location or a name. As a user, there are just strings to locate my images (maybe not everything thinks this way).

@sgotti I am not sure about the Type: Name vs Type: Location thing. At this point I can't see why Distribution is not enough to express the Name vs Location ?

My other point is since we already express the difference in the `Distribution` field, unless we combine Distribution: Appc with Distribution: ACIArchive, and Distribution: OCI with Distribution: OCIImageLayout. I don't see why a separate `Location` field is needed?

sgotti replied:
@yifan-gu the fields doesn't exists anywhere in the code, was just a way to explain the logical difference. I tried to explain the reasons in the above comment.

A user will only use a string (URI or an user friendly string).

If this table is a source of confusion I can drop it, but I'm sure this logical difference, from the developing perspective is important (see always above comment examples)

s-urbaniak commented on Documentation/devel/distribution.md:
```diff
@@ -0,0 +1,145 @@
+
+Since the OCI image layout can provide multiple images selectable by a ref there's the need to specify which ref to use in the archive distribution URI (see the above ref query parameter). Since distribution just covers one image it's not possible to import all the refs with just a distribution URI.
+
+TODO(sgotti). Define if oci-image-layout should internally handle both archive and directory based layouts or use two different distributions or a query parameter the explicitly define the layout (to avoid guessing if the url points to a single file or to a directory).
```
I vote for handling both the archive, and directory based layouts. To distuingish a file from a directory is easy from the implementation perspective with a `file://` scheme. Apparently `http(s)://` always implies "file".

sgotti replied:
> I vote for handling both the archive, and directory based layouts. To distuingish a file from a directory is easy from the implementation perspective with a file:// scheme

:+1: 

> Apparently http(s):// always implies "file".

I wasn't sure if https always implied a single file. My idea, also reading the spec and various issues, was that the oci image layout could also provide a directory layout over http (that will act as an easy way to implement a cas/ref based "registry" accessible by all the machines instead of using a docker registry).

To distinguish a file froma dir one could just choose to treat as files all the url with `.oci` extension. In addition the distribution could use a query parameter like `layout` (with values file/archive or dir) to avoid guessing.

lucab replied:
My understanding it that one of the point for this is to stop guessing, so it may make sense to have different types if they need to be handled separately. But I think we should postpone this decision to the actual point when we need to reference those.

s-urbaniak commented on common/distribution/dist.go:
```diff
@@ -0,0 +1,77 @@
+	// Returns a copy of the distribution URI
+	URI() *url.URL
+	// Returns the distribution type
+	Type() DistType
```
You are encoding the distribution type in two places, as `distribution.Docker`, and `DistTypeDocker`. I am not convinced you need this type information in two places.

s-urbaniak commented on common/distribution/dist.go:
```diff
@@ -0,0 +1,77 @@
+// A Distribution represnt the way to retrieve an image starting from an input
+// string.
+// It's identified by an URI with a specific schema
+type Distribution interface {
```
This interface does not define any behavior, just data.

s-urbaniak commented on common/distribution/dist.go:
```diff
@@ -0,0 +1,77 @@
+	DistScheme = "cimd"
+
+	// Distribution types
+	DistTypeAppc       DistType = "appc"
```
Rather than specifying strings here, why not go with concrete distribution instances, and let the `Distribution` interface implement only behavior, i.e `Compare(Distribution) bool`?

Compare https://github.com/opencontainers/image-spec/blob/master/schema/schema.go#L21.

sgotti commented:
As suggested, and to speed up things, I split the proposal (this PR) from the implementation (#3101).

@s-urbaniak thanks for your comments on the implementation! Due to the split they got lost in #3101, can you please copy them there?

s-urbaniak commented:
@sgotti copy-pasted my comments in #3101 :-)

lucab commented on Documentation/devel/distribution.md:
```diff
@@ -0,0 +1,147 @@
+
+**NOTE:** I'm not sure if distribution is the best name for this concept or it should be called with other names like "image reference". So I'd like to hear your thoughts.
```
I guess we can settle on calling them "Distribution Points". Current name looks fine enough, we just need to disambiguate it from typical "Linux distro" usage.

sgotti replied:
`Distribution Points` looks good. I was also thinking to just `Distribution URIs`.

lucab commented on Documentation/devel/distribution.md:
```diff
@@ -0,0 +1,147 @@
+* Some docker registries also provides OCI images.
+* Rkt can internally implement the docker distribution to natively fetch a docker/OCI image from a registry or an ACI converted by docker2aci.
+
+Before this concept was introduced, rkt used the ImageType concept, the ImageType was mapped to the input image string format and internally covered multiple concepts like distribution, transport and image type (hidden since all are now appc ACIs).
```
"(implicitly, since only appc was handled internally)"

lucab commented on Documentation/devel/distribution.md:
```diff
@@ -0,0 +1,147 @@
+The distribution can be logically split in two types:
+
+* Name based distribution
+* Location based distribution
```
I guess a better categorization here is indirect/direct. Name-based can perform additional indirect steps (eg. discovery, assembling, signature fetching) while location-based seems to suit direct go-to scenarios.

lucab commented on Documentation/devel/distribution.md:
```diff
@@ -0,0 +1,147 @@
+* Location based distribution
+
+Name based distribution can be considered the first class distribution type as it'll be the primarly used distribution type.
+Location based distribution can be thought as a sub-distribution for Name based distribution or a fast alternative (but not covering the Name based distribution feature) to directly fetch an image from a location.
```
I think this sentence and the one above only add more confusion. Primary usage depends on specific usecases. Again, I'd speak about direct and indirect resource referencing.

lucab commented on Documentation/devel/distribution.md:
```diff
@@ -0,0 +1,147 @@
+Name based distribution can be considered the first class distribution type as it'll be the primarly used distribution type.
+Location based distribution can be thought as a sub-distribution for Name based distribution or a fast alternative (but not covering the Name based distribution feature) to directly fetch an image from a location.
+
+A distribution is represented as an URI with the uri scheme as "cimd" and the opaque data and query parts as the distribution information (I started trying to use just URLs being an URI subset for location but distributions IMHO don't cleanly map to a resource locator but better to a name, [rfc3986](https://tools.ietf.org/html/rfc3986)) I'm calling them URIs instead of URNs because it's the sugge ...
```
I think we can settle on the "cimd" and expand here what it means. Also, I'm not sure what you mean with "the opaque data and query parts as the distribution information" (but I got the idea) so I think it may be better phrased.

lucab commented on Documentation/devel/distribution.md:
```diff
@@ -0,0 +1,147 @@
+
+A distribution is represented as an URI with the uri scheme as "cimd" and the opaque data and query parts as the distribution information (I started trying to use just URLs being an URI subset for location but distributions IMHO don't cleanly map to a resource locator but better to a name, [rfc3986](https://tools.ietf.org/html/rfc3986)) I'm calling them URIs instead of URNs because it's the sugge ...
+
+Every distribution starts with a common part: `cimd:DISTTYPE:V=uint32(VERSION):` where `cimd` is the container image distribution scheme, DISTTYPE is the distribution type, V=uint32(VERSION) is the distribution type format version.
```
`v=` before the version is actually lowercase everywhere else.

lucab commented on Documentation/devel/distribution.md:
```diff
@@ -0,0 +1,147 @@
+
+Appc defines a distribution using appc image discovery
+The format is: `cimd:appc:v=0:name?label01=....&label02=....`
+The scheme is "appc"
```
s/scheme/distribution type/

lucab commented on Documentation/devel/distribution.md:
```diff
@@ -0,0 +1,147 @@
+
+TODO(sgotti). Define if oci-image-layout should internally handle both archive and directory based layouts or use two different distributions or a query parameter the explicitly define the layout (to avoid guessing if the url points to a single file or to a directory).
+
+**Note** Considering [this OCI image spec README section](https://github.com/opencontainers/image-spec#running-an-oci-image), probably the final distribution format will be similar to the app distribution. So there's a need to distinguish their User Friendly string (prepending an appc: or oci: ?).
```
At that point, we'll need/have an external "User friendly distribution strings" format for OCI.

sgotti replied:
The main question is that from that README looks like the proposed user friendly distribution string for OCI clashes (it's identical) with the Appc one. So one idea will be to prefix the user friendly string with the distribution name (`appc:`, `oci:`).

lucab replied:
In this context it isn't a problem (we already have the distribution type for disambiguation). In the CLI interface (ie. receiving input from the user) yes, it would better if the two projects could specify their own specific schemes. But I fear we'll need get rid of a bit of heuristic (in the frontend).

lucab commented on Documentation/devel/distribution.md:
```diff
@@ -0,0 +1,147 @@
+
+The parsing and generation of user friendly string is done outside the distribution package.
+
+rkt uses two function to:
```
function**s**

sgotti commented:
@lucab Thanks for the review! I used the `Distribution Point` name and I hope to have addressed all your comments.

s-urbaniak commented:
@lucab ping for final pass

lucab commented:
I think this doc needs quite a bit of re-wording and polishing, but we are more interested in its content and the concepts it contains. Let's land it now and implement this. We will need to re-touch this anyway to show how to use them in practice. LGTM.
